Network/Converter/PacketConverter.cs

```diff
@@ -27,321 +28,680 @@
-        private object ReadPrimitiveFromStream(PropertyInfo propertyInfo, BinaryReader binaryReader)
+        /// <param name="propertyInfo">
+        /// The <see cref="PropertyInfo"/> to deserialise from the given
+        /// <see cref="BinaryReader"/>s underlying <see cref="MemoryStream"/>.
+        /// </param>
+        /// <param name="binaryReader">
+        /// The <see cref="BinaryReader"/> from whose underlying <see cref="MemoryStream"/>
+        /// to deserialise the primitive.
+        /// </param>
+        /// <returns>
+        /// The primitive object that was deserialised from the <see cref="MemoryStream"/>.
+        /// </returns>
+        /// <remarks>
+        /// This method can return 'null' if the <see cref="NetworkObjectState"/>
+        /// is <see cref="NetworkObjectState.Null"/>.
+        /// </remarks>
+        private object ReadPrimitiveFromStream(
+            PropertyInfo propertyInfo, BinaryReader binaryReader)
         {
-            ObjectState objectState = (ObjectState)binaryReader.ReadByte();
+            NetworkObjectState networkObjectState =
+                (NetworkObjectState)binaryReader.ReadByte();
 
-            if(objectState == ObjectState.NOT_NULL)
-                return ReadPrimitiveFromStream(propertyInfo.PropertyType, binaryReader);
+            if (networkObjectState == NetworkObjectState.NotNull)
+            {
+                return ReadPrimitiveFromStream(
+                    propertyInfo.PropertyType, binaryReader);
+            }
 
             return null;
         }
 
         /// <summary>
-        /// Reads a primitive from a stream.
+        /// Reads a primitive type from the given <see cref="BinaryReader"/>s
+        /// underlying <see cref="MemoryStream"/> and returns it.
         /// </summary>
-        /// <param name="type">The type.</param>
-        /// <param name="binaryReader">The binary reader.</param>
-        /// <returns>System.Object.</returns>
+        /// <param name="type">
+        /// The <see cref="Type"/> of the primitive to read from the given
+        /// <see cref="BinaryReader"/>s underlying <see cref="MemoryStream"/>.
+        /// </param>
+        /// <param name="binaryReader">
+        /// The <see cref="BinaryReader"/> from whose underlying <see cref="MemoryStream"/>
+        /// to read the primitive.
+        /// </param>
+        /// <returns>
+        /// The primitive that was read from the given <see cref="BinaryReader"/>s
+        /// underlying <see cref="MemoryStream"/>.
+        /// </returns>
+        /// <exception cref="NotSupportedException">
+        /// Thrown whenever a <see cref="Type"/> is passed to this method that
+        /// is not a primitive.
+        /// </exception>
         private object ReadPrimitiveFromStream(Type type, BinaryReader binaryReader)
         {
-            if (type.Equals(typeof(String)))
+            #region Reading Primitives From Stream
+
+            if (type == typeof(String))
```

Comment:
why removing all the "else". This method doesn't support several types (so if, if, if would make sense) It supports only one. Also, this implementation would evaluate each if, resulting in bad performance

Reply:
Did not know that it could decrease performance. Could you explain this ( i will change it back dw)

Reply:
Imagine the first if is true. All other ifs still will be evaluated, even thought they can't be true in any case.

Reply:
surely thats the same anyway? even with an 'else if', during debugging and stepping through you can see that the condition is being evaluated. Since the branch returns right after the 'if', surely no other condition *can* be evaluated?